src/tss2-tcti/tcti-libtpms.c

```diff
@@ -0,0 +1,969 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*******************************************************************************
+ * Copyright 2019, Fraunhofer SIT, Infineon Technologies AG, Intel Corporation
+ * All rights reserved.
+ ******************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <inttypes.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <dlfcn.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/syscall.h>
+#include <netinet/in.h>
+#include <pthread.h>
+#include "tss2_tcti_libtpms.h"
+
+#include "tcti-libtpms.h"
+#include "tcti-common.h"
+#define LOGMODULE tcti
+#include "util/log.h"
+
+/*
+ * libtpms API calls need to be wrapped. We set the current active TCTI module
+ * for this thread. This is needed because libtpms may call callbacks and these
+ * need to know which TCTI context they have to operate on.
+ *
+ * This macro assumes that TSS2_RC rc and int ret are declared. Jumps to
+ * fail_label on error. In this case, rc contains the respective error code.
+ */
+#define LIBTPMS_API_CALL(fail_label, tcti_libtpms, function, ...) \
+    rc = tcti_libtpms_set_current_tcti(tcti_libtpms); \
+    if (rc != TPM2_RC_SUCCESS)  { \
+        goto fail_label; \
+    } \
+    ret = tcti_libtpms->function(__VA_ARGS__); \
+    if (ret != TPM_SUCCESS) { \
+        LOG_ERROR("libtpms function " #function "() failed with return code 0x%" PRIx32, ret); \
+        rc = TSS2_TCTI_RC_GENERAL_FAILURE; \
+        goto fail_label; \
+    } \
+    rc = tcti_libtpms_clear_current_tcti(); \
+    if (rc != TPM2_RC_SUCCESS)  { \
+        goto fail_label; \
+    }
+
+typedef struct {
+    pid_t thread_id;
+    TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms;
+} tcti_instance;
+
+/*
+ * List which associates each registered thread with the currently active
+ * libtpms tcti.
+ */
+static tcti_instance current_libtpms_tctis[64] = {0};
+pthread_mutex_t current_libtpms_tctis_lock;
```

Comment:
I still don't quite get the idea why do we need to map active tcti instances to threads. Is it because the callbacks are called asynchronously? Couldn't find any info about that in the libtpms docs.
Perhaps some general comment on why this is needed would be good here.

Reply:
The problem is this: you cannot implement localities without registering at least the locality callback. The callback headers, however, provide no way for passing down _any_ data. Even the `tpm_number` parameter is hard-coded to `0` on libtpms side. This is a problem if we have multiple libtpms-instances - there is no way to distinguish different contexts from within the callback.

My solution is having a global variable and setting the "currently active" libtpms-context before every libtpms API call and clearing it again afterwards. This is why I have the wrapper macro `LIBTPMS_API_CALL()`. This only works because libtpms does _not_ call the callbacks asynchronously.

Of course you could have multiple threads. So in fact, I track the "currently active" tcti _per thread_.

Reply:
Since you need a map from thread id to context to get the proper context on libtpms callbacks, you could avoid the complexity by using thread local storage. Keyword __thread has been supported for a long time, and makes it even easier.

```C
__thread THREAD_CONTEXT state;
```

Or you can use pthread local storage API, see https://linux.die.net/man/3/pthread_setspecific
example here: https://gist.github.com/jbwyme/9479813

Then you won't need to manage a global map.

Reply:
Dang it, if only I had known that earlier... I'll rewrite that bit.

I intentionally did not use pthread (except for the mutex) because I wanted to support non-pthread (i.e. syscall) threads. I'm not too sure if that makes a difference, though. Guess I'll use `__thread ` now, anyway.

Learnt something today, thanks!